### PR TITLE
WIP: node/http: dumpzone

### PR DESCRIPTION
**Pull request by tynes**
Experimenting with dumping the entire Handshake tree as a zonefile. This would allow a traditional DNS server to serve the Handshake zone. A tradeoff in security for scalability.

We need to figure out how often this would be called and ensure that its performant enough that it will not cause any problems with staying in consensus. It needs to be tested on a very large tree, planning on testing it on the current testnet.

Closes https://github.com/handshake-org/hsd/issues/152

### CODE REVIEW

File: lib/node/http.js [hunk 1]
```diff
@@ -404,6 +408,59 @@ class HTTP extends Server {
...
+            if (seen === 0)
+              res.write(`"${entry}"`);
+            else
+              res.write(`,"${entry}"`);
+
+            seen++;
+          }
+        }
+      }
+
```
Comment (tynes) [inline lib/node/http.js]: Log number of records seen here

**Comment by jacobhaven** [comment]
@tynes @chjj Do we know how this will change when RR types are removed for #125 ?

Also, to build a working public resolver, it would be really nice to support iterating every HS name that has changed since a given tree interval (potentially ALL HS names) and dumping them all as zones for a normal DNS NS to serve. We've been working towards this with @mikedamm

**Comment by tynes** [comment]
@jacobhaven 

The `Resource` object should handle the `version` internally and not leak that to the consumer of the `Resource` object. If the `Resource` has a `version` that does not support a particular resource record type, it should not create any resource records in the `toDNS` method. In the case of the code in this PR, `msg.records()` should return an empty iterator: 

```javascript
const msg = resource.toDNS(fqdn, rrtype);

for (const rr of msg.records()) {
  const entry = rr.toString();
}
```

Note that this PR is naive in the sense that it uses brute force, there could be better ways to do it. Also see https://github.com/handshake-org/hsd/issues/125#issuecomment-547652205 as it is impossible to truly remove v0 serialization unless serialization versions become part of consensus.

**Comment by jacobhaven** [comment]
@tynes I don't think doing this in a brute force way is really even that bad, as it's only used every 36 blocks. The problem you also need some way to enumerate all names that have changed since you last synced. It's not clear to me how to do that.

You can "truly remove" and old version of DNS serialization by just no longer resolving it. I doubt that simultaneously supporting multiple types of DNS Zones would be a good idea. Each resolver can just deprecate old ones as it sees fit.

**Comment by tynes** [comment]
> The problem you also need some way to enumerate all names that have changed since you last synced.

This is a very interesting problem to solve, I think a closer inspection of the database would be helpful in solving this:

- https://github.com/handshake-org/urkel

Note that there is also an implementation written in Golang by Oasis Labs

- https://github.com/oasislabs/oasis-core/tree/master/go/storage/mkvs/urkel

----

> You can "truly remove" and old version of DNS serialization by just no longer resolving it.

This is an inherently subjective removal. The only "objective truth" is found through Nakamoto Consensus + Proof of Work, assuming minority byzantine miners (51% attack) and peers (eclipse attack). Whatever is on chain is on chain. Removing the ability to resolve (encode/decode the on chain data) a particular `Resource` version would effectively "brick" the usage of the name. The price of exit is pretty low, assuming that users can run their own full nodes or its relatively easy for new service providers to run a set of full nodes. Users will exit to the resolvers that best serve them.

**Comment by pinheadmz** [comment]
Closing this to focus development on #534 thank you @tynes ! Handsome !

**Pull request closed without merging**